Cubical/Data/Vec/NAry.agda

```diff
@@ -0,0 +1,50 @@
+{-# OPTIONS --cubical --safe #-}
+module Cubical.Data.Vec.NAry where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+
+open import Cubical.Data.Nat
+open import Cubical.Data.Vec.Base
+
+private
+  variable
+    ℓ ℓ' : Level
+    A : Type ℓ
+    B : Type ℓ'
+
+nAryLevel : Level → Level → ℕ → Level
+nAryLevel ℓ₁ ℓ₂ zero    = ℓ₂
+nAryLevel ℓ₁ ℓ₂ (suc n) = ℓ-max ℓ₁ (nAryLevel ℓ₁ ℓ₂ n)
+
+nAryOp : (n : ℕ) → Type ℓ → Type ℓ' → Type (nAryLevel ℓ ℓ' n)
+nAryOp zero A B    = B
+nAryOp (suc n) A B = A → nAryOp n A B
+
+
+_$ⁿ_ : ∀ {n} → nAryOp n A B → (Vec A n → B)
+f $ⁿ []       = f
+f $ⁿ (x ∷ xs) = f x $ⁿ xs
+
+curryⁿ : ∀ {n} → (Vec A n → B) → nAryOp n A B
+curryⁿ {n = zero}  f   = f []
+curryⁿ {n = suc n} f x = curryⁿ (λ xs → f (x ∷ xs))
+
+$ⁿ-curryⁿ : ∀ {n} (f : Vec A n → B) → _$ⁿ_ (curryⁿ f) ≡ f
+$ⁿ-curryⁿ {n = zero} f  = funExt λ { [] → refl }
+$ⁿ-curryⁿ {n = suc n} f = funExt λ { (x ∷ xs) i → $ⁿ-curryⁿ {n = n} (λ ys → f (x ∷ ys)) i xs}
+
+curryⁿ-$ⁿ : ∀ {n} (f : nAryOp {ℓ = ℓ} {ℓ' = ℓ'} n A B) → curryⁿ {A = A} {B = B} (_$ⁿ_ f) ≡ f
+curryⁿ-$ⁿ {n = zero} f  = refl
+curryⁿ-$ⁿ {n = suc n} f = funExt λ x → curryⁿ-$ⁿ {n = n} (f x)
+
+nAryOp≃VecFun : ∀ {n} → nAryOp n A B ≃ (Vec A n → B)
+nAryOp≃VecFun {n = n} = isoToEquiv f
+  where
+  f : Iso (nAryOp n A B) (Vec A n → B)
```

Comment:
I've noticed this pattern of creating a where clause for the purpose of constructing an `Iso` via copatterns. Would it make sense to use the copattern version of [pattern matching lambda](https://agda.readthedocs.io/en/v2.6.1/language/lambda-abstraction.html) like we do [for `Conats`](https://github.com/agda/cubical/blob/4c3a3e2e06c86bce9bdb70d8e863e5ad73b97eb7/Cubical/Codata/Conat/Base.agda#L47)?

Reply:
I guess that should work, but I think it's easier to read this way. I have no strong opinions about this though so if someone wants to rewrite these please do

Reply:
Pattern matching lambdas have the downside that they will be printed in full most of the time, and otherwise give you an uninformative generated name.

Reply:
I see, that seems like an argument to keep things the way they are